File: wasm/parity.mjs
// WebAssembly parity harness.
//
// Loads the js/wasm build of the signing core and replays every golden vector
// through it, asserting that the preimage, the payload and the final signed
// entry are byte-identical to the recorded values. The vectors are what the
// native Go build and @stellar/stellar-sdk already agree on, so matching them
// here is the evidence that the wasm build did not change the bytes.
//
// The harness talks to the module exactly as a browser would: it passes base64
// XDR strings in and reads base64 XDR strings out. A source-account entry has
// no preimage (the transaction envelope covers it), so it is exercised through
// the pass-through path instead and reported as skipped, loudly, with the
// reason.
//
// Usage: node wasm/parity.mjs [--wasm wasm/dist/soroauth.wasm] [--vectors testdata/vectors]
// Exit status is 0 only when every vector matched and at least one was checked.

import { createHash } from "node:crypto";
import { createRequire } from "node:module";
import { readdirSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(HERE, "..");

// A fixed seed for the source-account pass-through probe. It is never used to
// sign anything -- the pass-through returns the entry before a signer is
// consulted -- it just has to be a valid 32-byte seed.
const PROBE_SEED = "00".repeat(32);

function parseArgs(argv) {
  let wasm = join(HERE, "dist", "soroauth.wasm");
  let vectors = join(ROOT, "testdata", "vectors");
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--wasm") wasm = resolve(argv[++i]);
    else if (argv[i] === "--vectors") vectors = resolve(argv[++i]);
    else throw new Error(`unknown argument: ${argv[i]}`);
  }
  return { wasm, vectors };
}

// loadRuntime installs the globals wasm_exec.js expects and evaluates it,
// leaving globalThis.Go defined. The shim is copied next to the wasm by
// wasm/build.sh so the two always match.
async function loadRuntime(wasmPath) {
  const require = createRequire(import.meta.url);
  globalThis.require = require;
  globalThis.fs = require("node:fs");
  globalThis.path = require("node:path");

  const shim = join(dirname(wasmPath), "wasm_exec.js");
  let source;
  try {
    source = readFileSync(shim, "utf8");
  } catch {
    throw new Error(
      `missing ${shim}; run wasm/build.sh to produce the module and its runtime shim`,
    );
  }
  // The shim is a plain IIFE that defines globalThis.Go; evaluating it as a
  // script (not as a module) matches how a browser script tag would load it.
  (0, eval)(source);

  const go = new globalThis.Go();
  go.argv = ["soroauth.wasm"];
  go.env = {};

  const bytes = readFileSync(wasmPath);
  const { instance } = await WebAssembly.instantiate(bytes, go.importObject);

  // The Go program blocks in select{} after it installs its bindings, so this
  // promise never settles. Do not await it.
  go.run(instance);

  // Wait for main to install the bindings. Bounded so a broken module fails
  // loudly instead of hanging the job.
  for (let i = 0; i < 1000; i++) {
    if (globalThis.soroauth) return globalThis.soroauth;
    await new Promise((r) => setTimeout(r, 10));
  }
  throw new Error("timed out waiting for globalThis.soroauth from the wasm module");
}

function sha256Hex(label) {
  return createHash("sha256").update(label).digest("hex");
}

// call invokes a binding and throws on a {ok:false} envelope, the same way the
// TypeScript wrapper does.
function call(api, name, ...args) {
  const fn = api[name];
  if (typeof fn !== "function") throw new Error(`module does not export ${name}`);
  const result = fn(...args);
  if (!result || result.ok !== true) {
    throw new Error(`${name}: ${(result && result.error) || "unknown error"}`);
  }
  return result;
}

function loadVectors(dir) {
  const names = readdirSync(dir)
    .filter((n) => n.endsWith(".json"))
    .sort();
  if (names.length === 0) throw new Error(`no vectors found in ${dir}`);
  return names.map((n) => JSON.parse(readFileSync(join(dir, n), "utf8")));
}

function vectorsMatch(api, vector) {
  // A source-account entry has no preimage. Exercise the pass-through instead:
  // authorizeWithSeed must hand the entry back byte-identical.
  let preimage;
  try {
    preimage = call(api, "preimage", vector.unsigned_entry_xdr, vector.valid_until_ledger, vector.network_passphrase);
  } catch (err) {
    if (!/source.account/i.test(err.message)) throw err;
    const passed = call(
      api, "authorizeWithSeed",
      vector.unsigned_entry_xdr, vector.valid_until_ledger,
      vector.network_passphrase, PROBE_SEED, "",
    ).entryB64;
    if (passed !== vector.unsigned_entry_xdr) {
      return { status: "mismatch", detail: "source-account pass-through changed the entry" };
    }
    return {
      status: "skip",
      detail: `no preimage: credential_type=source_account (pass-through verified): ${err.message}`,
    };
  }

  if (preimage.preimageXdr !== vector.preimage_xdr) {
    return { status: "mismatch", detail: "preimage differs" };
  }
  if (preimage.payloadHex !== vector.payload_hex) {
    return { status: "mismatch", detail: "payload differs" };
  }

  // The standalone payload entry point must agree with the one computed from
  // the entry; a remote signer that only receives the preimage depends on it.
  const standalone = call(api, "payload", preimage.preimageXdr).payloadHex;
  if (standalone !== vector.payload_hex) {
    return { status: "mismatch", detail: "payload(preimage) differs from recorded payload" };
  }

  // Replay every signing step, exactly as golden_test.go does.
  let current = vector.unsigned_entry_xdr;
  for (const step of vector.steps) {
    current = call(
      api, "authorizeWithSeed",
      current, vector.valid_until_ledger, vector.network_passphrase,
      sha256Hex(step.signer_label), step.for_address ?? "",
    ).entryB64;
  }
  if (current !== vector.signed_entry_xdr) {
    return { status: "mismatch", detail: "signed entry differs" };
  }

  return { status: "match", detail: "" };
}

async function main() {
  const { wasm, vectors } = parseArgs(process.argv.slice(2));
  console.log("WebAssembly parity harness");
  console.log(`wasm:    ${wasm}`);
  console.log(`vectors: ${vectors}`);

  const api = await loadRuntime(wasm);
  const cases = loadVectors(vectors);

  let matched = 0;
  let mismatched = 0;
  let skipped = 0;

  for (const vector of cases) {
    const result = vectorsMatch(api, vector);
    if (result.status === "match") {
      matched++;
      console.log(`  ok    ${vector.name}`);
    } else if (result.status === "mismatch") {
      mismatched++;
      console.error(`  FAIL  ${vector.name}: ${result.detail}`);
    } else {
      skipped++;
      console.error(`  skip  ${vector.name}: ${result.detail}`);
    }
  }

  console.log(
    `\nWebAssembly parity: ${matched} matched, ${skipped} skipped, ${mismatched} mismatched (of ${cases.length})`,
  );

  if (mismatched > 0) {
    console.error(
      "\nThe wasm build does not reproduce the native golden vectors. Do not edit a vector; fix the module.",
    );
    return 1;
  }
  if (matched === 0) {
    console.error("nothing was actually checked; refusing to report success");
    return 1;
  }
  return 0;
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });

File: testdata/parity-python/parity.py
#!/usr/bin/env python3
"""Python stellar-sdk parity harness for the soroauth golden vectors.

The golden vectors prove soroauth agrees with ``@stellar/stellar-sdk`` byte for
byte. They cannot prove that agreement is *correct*: a bug committed by both the
Go library and the JS reference would be baked into the vectors and every test
would keep passing. Recomputing the preimage and the payload with a third,
independently maintained implementation is the cheapest way to catch that class
of bug.

This harness reads every ``testdata/vectors/*.json`` the JS generator wrote,
rebuilds the ``HashIDPreimage`` from the recorded ``unsigned_entry_xdr``,
``valid_until_ledger`` and ``network_passphrase`` with the Python
``stellar-sdk``, and asserts that:

* the marshalled preimage equals the recorded ``preimage_xdr``;
* ``sha256(preimage)`` equals the recorded ``payload_hex``.

It never writes to a vector. A disagreement is a failure to investigate, not a
value to update: see the accompanying README.

Cases the Python SDK cannot express are skipped **loudly** -- named on stderr
and counted in the summary -- and a run that checked nothing is a failure, so
the harness cannot pass by doing nothing.

Pinned reference: ``stellar-sdk==16.1.0`` (see ``requirements.txt``). The
harness reads the version that is actually installed and refuses to run against
anything else, the same way ``testdata/gen/gen.mjs`` does for its JS SDK.

Run it with:

    python3 testdata/parity-python/parity.py

Exit status is 0 only when every non-skipped vector matched and at least one
vector was checked.
"""

from __future__ import annotations

import argparse
import dataclasses
import glob
import json
import os
import sys
from importlib.metadata import PackageNotFoundError
from importlib.metadata import version as installed_version
from pathlib import Path

# The exact Python SDK this harness is pinned to. Vectors are only meaningful
# when they are compared against a known reference build, so an installed
# version other than this one is refused rather than compared.
REQUIRED_SDK_VERSION = "16.1.0"

# Repository layout: this file lives in testdata/parity-python/, next to
# testdata/vectors/.
HERE = Path(__file__).resolve().parent
DEFAULT_VECTORS_DIR = HERE.parent / "vectors"

# SorobanCredentialsType discriminants, named for readable output.
_CREDENTIAL_TYPES = {
    0: "source_account",
    1: "address",
    2: "address_v2",
    3: "address_with_delegates",
}


def credential_type_name(value: int) -> str:
    """Name a Soroban credential arm for human-readable output."""
    return _CREDENTIAL_TYPES.get(value, f"unknown({value})")


# ---------------------------------------------------------------------------
# Result model
# ---------------------------------------------------------------------------


@dataclasses.dataclass(frozen=True)
class CaseResult:
    """The outcome of checking one vector."""

    name: str
    status: str  # "match", "mismatch" or "skip"
    detail: str = ""

    @property
    def matched(self) -> bool:
        return self.status == "match"

    @property
    def mismatched(self) -> bool:
        return self.status == "mismatch"

    @property
    def skipped(self) -> bool:
        return self.status == "skip"


# ---------------------------------------------------------------------------
# Dependency check
# ---------------------------------------------------------------------------


def require_pinned_sdk() -> str:
    """Return the pinned SDK version, or exit if the installed one differs.

    Reading the version from the installed distribution -- rather than trusting
    a constant -- is what makes the pin meaningful: a stale virtualenv with a
    different SDK is detected here instead of silently producing a second
    opinion nobody can attribute to a version.
    """
    try:
        found = installed_version("stellar-sdk")
    except PackageNotFoundError:
        sys.stderr.write(
            "parity harness: the Python 'stellar-sdk' package is not installed.\n"
            "Install the pinned reference with:\n"
            f"  pip install -r {HERE / 'requirements.txt'}\n"
        )
        raise SystemExit(2)

    if found != REQUIRED_SDK_VERSION:
        sys.stderr.write(
            f"parity harness: installed stellar-sdk is {found}, this harness is "
            f"pinned to {REQUIRED_SDK_VERSION}.\n"
            f"Install the pinned reference with:\n"
            f"  pip install -r {HERE / 'requirements.txt'}\n"
        )
        raise SystemExit(2)

    return found


# ---------------------------------------------------------------------------
# Per-vector verification
# ---------------------------------------------------------------------------


def check_vector(path: Path) -> CaseResult:
    """Recompute one vector's preimage and payload and compare to the record.

    ``path`` is the JSON file. An unrepresentable case comes back as a ``skip``
    result, and a genuine disagreement as a ``mismatch``; neither raises.
    """
    # Imported here so ``require_pinned_sdk`` can print its friendly message
    # before a missing install would raise ImportError.
    import stellar_sdk.xdr as xdr
    from stellar_sdk.auth import (
        authorization_payload_hash,
        build_authorization_preimage,
    )

    with path.open(encoding="utf-8") as handle:
        vector = json.load(handle)

    name = vector.get("name") or path.name
    entry = xdr.SorobanAuthorizationEntry.from_xdr(vector["unsigned_entry_xdr"])
    cred_type = int(entry.credentials.type)

    if not vector.get("preimage_xdr"):
        # A vector with an empty recorded preimage is one the recording side
        # decided has no preimage at all -- a source-account entry, which the
        # transaction envelope authenticates instead. There is nothing to
        # recompute, so this is skipped, loudly, by name and reason.
        return CaseResult(
            name=name,
            status="skip",
            detail=(
                "no recorded preimage: "
                f"credential_type={credential_type_name(cred_type)}; cases with "
                "no preimage cannot be recomputed"
            ),
        )

    try:
        preimage = build_authorization_preimage(
            entry,
            vector["valid_until_ledger"],
            vector["network_passphrase"],
        )
    except ValueError as exc:
        # The Python SDK actively refuses an arm it cannot express. That is a
        # skip, not a pass, and the SDK's own message is the reason.
        return CaseResult(
            name=name,
            status="skip",
            detail=(
                f"credential_type={credential_type_name(cred_type)}: Python "
                f"stellar-sdk cannot express this case: {exc}"
            ),
        )

    got_preimage = preimage.to_xdr()
    got_payload = authorization_payload_hash(preimage).hex()
    want_preimage = vector.get("preimage_xdr", "")
    want_payload = vector.get("payload_hex", "")

    problems: list[str] = []
    if got_preimage != want_preimage:
        problems.append(
            f"preimage differs\n    want {want_preimage}\n     got {got_preimage}"
        )
    if got_payload != want_payload:
        problems.append(f"payload differs\n    want {want_payload}\n     got {got_payload}")

    if problems:
        return CaseResult(name=name, status="mismatch", detail="; ".join(problems))

    return CaseResult(name=name, status="match")


# ---------------------------------------------------------------------------
# Driver
# ---------------------------------------------------------------------------


def run(vectors_dir: Path) -> list[CaseResult]:
    """Check every vector in ``vectors_dir`` and return the per-vector results."""
    paths = sorted(Path(p) for p in glob.glob(os.fspath(vectors_dir / "*.json")))
    if not paths:
        raise SystemExit(
            f"parity harness: no vectors found in {vectors_dir}; "
            "run `cd testdata/gen && npm ci && node gen.mjs` to generate them"
        )
    return [check_vector(path) for path in paths]


def report(results: list[CaseResult], out=sys.stdout, err=sys.stderr) -> int:
    """Print a human-readable report and return the process exit code."""
    matched = [r for r in results if r.matched]
    mismatched = [r for r in results if r.mismatched]
    skipped = [r for r in results if r.skipped]

    for result in results:
        if result.matched:
            print(f"  ok    {result.name}", file=out)
        elif result.skipped:
            # Skips go to stderr so they cannot be mistaken for the "ok" lines,
            # and they name the case and the reason.
            print(f"  skip  {result.name}: {result.detail}", file=err)

    for result in mismatched:
        print(f"  FAIL  {result.name}: {result.detail}", file=err)

    print(
        f"\nparity harness: {len(matched)} matched, {len(skipped)} skipped, "
        f"{len(mismatched)} mismatched (of {len(results)})",
        file=out,
    )

    if mismatched:
        print(
            "\nDo not edit a vector to make this pass. A disagreement here means "
            "one of the two implementations is wrong; open an issue with the "
            "protocol reference (CAP-71-01) and investigate.",
            file=err,
        )
        return 1

    # A run that checked nothing is not a pass. This is the guard against the
    # harness silently degrading to a no-op if every vector becomes skippable.
    if not matched:
        print(
            "parity harness: nothing was actually checked; refusing to report success",
            file=err,
        )
        return 1

    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description=(
            "Recompute the golden vectors' preimages and payloads with the "
            "Python stellar-sdk and compare them against the recorded values."
        )
    )
    parser.add_argument(
        "--vectors",
        type=Path,
        default=DEFAULT_VECTORS_DIR,
        help=(
            "directory holding the golden vector JSON files "
            f"(default: {DEFAULT_VECTORS_DIR})"
        ),
    )
    args = parser.parse_args(argv)

    sdk_version = require_pinned_sdk()
    print(f"Python stellar-sdk {sdk_version} parity harness")
    print(f"vectors: {args.vectors}")

    results = run(args.vectors)
    return report(results)


if __name__ == "__main__":
    raise SystemExit(main())
